### ISSUE TITLE
Offline support inquiry

### ISSUE THREAD

**Issue by harryworld** [caiyue1993/IceCream#7]
CloudKit recommends to track local changes for offline work.

**Comment by caiyue1993** [comment]
Could you please gimme more details about "track local changes for offline work"? That will help me to locate the issue. For example you can give a example on it.

**Comment by harryworld** [comment]
Thanks for the reply.

I built a simple task list app using IceCream, and turned on Airplane intentionally. Once the tasks are updated, Realm notification is triggered and attempt to push to CloudKit. This will fail since network is disconnected.

Then I restart the app, turn off Airplane mode. Since I did not update the same task again, the data is not synced to CloudKit.

**Comment by caiyue1993** [comment]
Well done, you have described the issue clearly. 

Since the CKOperation has a property `isLongLived` in the [CKOperationConfiguration](https://developer.apple.com/documentation/cloudkit/ckoperationconfiguration) and I have already turned this property to true. So theoretically CloudKit api handles the offline support for us. 

I also followed the steps you offer myself when I develop IceCream and everything was fine. So I may have 2 questions for you:
1. Are you testing on Simulator or real device(recommended)?
2. What's your iOS version?

**Comment by harryworld** [comment]
I tested on iPhone 6 Plus with iOS 11.1 (15B93)

**Comment by insidegui** [comment]
@caiyue1993 CloudKit does not handle the case where an operation fails automatically. You have to handle the errors returned by its operations and react accordingly. I think one approach would be to track which records have been uploaded successfully and which ones have failed to upload and try to upload them again once the network is reachable. You'll also need to handle possible merge conflicts.

For the latter, I think IceCream could have a simple merge conflict handler that uses the most recent record, but it should allow the user of the framework to provide a custom merge conflict handler for each record type.

**Comment by harryworld** [comment]
@insidegui I agree. From [WWDC2016](https://developer.apple.com/videos/play/wwdc2016/231/?time=1650), long-lived operation only allows to replay a successful operation happened. In the scenario of AirPlane, it is an error indicating the operation is not done.

I wonder what would be nice ways to track the records to be uploaded. I have some ideas here:

1. Extra fields in Realm Object (e.g. isDirty? isUploading?)
2. Extra tables in Realm (e.g. Object Type and Primary Key?)
3. 3rd-Party dependency (we used PINCache before)

Regarding to merge conflict, I can see a need for using the default save policy `ifServerRecordUnchanged`, which will have a need to resolve conflict.

**Comment by caiyue1993** [comment]
@insidegui @harryworld Thanks a lot for your guys' advice. 

So, before we dive deeper to this problem I wanna ensure some playbacks from you @harryworld. In your reply, you mentioned: 

"Once the tasks are updated, Realm notification is triggered and attempt to push to CloudKit. This will fail since network is disconnected.**Then I restart the app**, turn off Airplane mode. Since I did not update the same task again, the data is not synced to CloudKit."

What do you mean by "Then I restart the app"? Do you mean by killing the app in the background and open the app again, or, just press the home button and bring the app to foreground again? Which one? 

Your answer will help me not go to the wrong way.

**Comment by insidegui** [comment]
I think the best approach would be to keep a separate Realm where IceCream can store metadata about the syncing. Adding fields to every model would get messy.

**Comment by harryworld** [comment]
@caiyue1993 it’s a kill app approach. Since I’m running Xcode build, so it’s just a re-run

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by caiyue1993** [comment]
OK. I have already known this issue. Once again, in my part I followed your @harryworld steps, the left-over data still uploaded to CloudKit successfully. I know the issue may still stick around in your side. Don't worry, I believe there should be a way to easily fix it.  

According to the [CloudKit Best Practices](https://developer.apple.com/videos/play/wwdc2016/231/?time=1650):

> What these operations(Long-lived operation specified) let you do is, once you mark them long-lived, the system will execute that operation on your apps behalf whether or not your app sticks around. 

And I have already added a `resumeLongLivedOperationIfPossible()` method in the initialization process of SyncEngine, which will execute when app launches. 

So if you do these following steps:
1.  turn on Airplane mode
2. update a data
3. exit app(kill in the background)
4. relaunch app

At this time, SyncEngine will execute `resumeLongLivedOperationIfPossible`(whether I turn off airplane mode before or after step 4) and replay these longLived operation and continue to execute it.

And I highly recommend looking at the [CKOperation Doc](https://developer.apple.com/documentation/cloudkit/ckoperation). In the bottom of the page, it shows a complete routine of long-lived operation. 

So could you please double-check the documentation and WWDC video? And we could discuss and practice more on that.

**Comment by harryworld** [comment]
Thanks for the detail. Perhaps Airplane mode is not as clear (since in iOS 11, turning on Airplane mode may still leave wifi on). Instead, I turned off all connections this time and uses the IceCream example app to reproduce the issue.

1. Run `IceCream Example` project in a clean container
2. (Connected) Add Dogs 1, 2
3. (Disconnected) Add Dog 3
4. (Disconnected) Restart the app
5. (Connected) Add Dog 4

In this case, Dog 3 will not be available in CloudKit.

The problem is not related to Long-Lived operations, but rather an operation will error (e.g. connection error)

**Comment by caiyue1993** [comment]
Thanks @harryworld for the patience. I tried a lot of times and I have successfully reproduced your problem and I **totally understand** what you're talking about. All of the operation errors are handled in SyncEngine already.

I just want to hold on a second and I also want you to have a try. The key is the step 4: "(Disconnected) Restart the app". When I restart the app by running Xcode again, the Dog 3 will not be tracked. But when I manually force exit my app(in the background), the Dog 3 synced to the iCloud successfully.

Would you mind giving it a try in the latter method?  Hoping to receive your response. :)

**Comment by caiyue1993** [comment]
Many thanks for the suggestions! @insidegui 

I'll dig this issue deeper and find a simplest implementation.

**Comment by harryworld** [comment]
@caiyue1993  Just tried, the missing Dog is still missing after force exit.

**Comment by caiyue1993** [comment]
@harryworld OK. So that's weird but I have already known your situation. And I'm gonna organize the information you and @insidegui have offered. And I'm gonna do further search and tell you the result. 

This issue will keep open. If anyone else has encountered this problem, please comment below and let me know.

**Comment by aaronpearce** [comment]
Hitting a similar issue. I'm trying to use IceCream behind a IAP so enable SyncEngine after some objects could exist, and sometimes these don't sync correctly. 

Especially if I turn sync off on one device, make a new record on that device, then set SyncEngine back up again by turning sync on. The second device will not see that record.

**Comment by aehlke** [comment]
@aaronpearce Have you tried using soft-deletion instead? Then you can do conflict resolution.

**Comment by aaronpearce** [comment]
Soft deletion is how IceCream works. You mark something as “isDeleted”

**Comment by aaronpearce** [comment]
Possibly the system needs to use a deleted at date instead so it can compare.

**Comment by insidegui** [comment]
@aaronpearce Keep in mind the App Review guidelines don't allow you to charge for built-in system functionality (CloudKit in this case).

**Comment by aaronpearce** [comment]
@insidegui Realised that and launched as a paid up front app.. Amusingly plenty of other apps already do charge via IAP/subscription for CloudKit syncing.

**Comment by aehlke** [comment]
You can likely charge for syncing if it's part of some (even slightly) broader feature that you pay for, so that it's not the syncing itself in isolation that's being paid for.

**Comment by insidegui** [comment]
> Amusingly plenty of other apps already do charge via IAP/subscription for CloudKit syncing.

Yes, that is unfortunate.

**Comment by caiyue1993** [comment]
Close this one because of no more activities. If you're facing the offline problem, please create a new issue ticket.
